util/timer/src/timer.rs

```diff
@@ -0,0 +1,687 @@
+        let mut scheduler = self.inner.lock();
+        let result = scheduler.cancel(timer_id)?;
+        self.condvar.notify_all();
+        Ok(result)
+    }
+
+    fn run(&self, worker_queue: &WorkerQueue) {
+        let mut scheduler = self.inner.lock();
+        while !scheduler.stop {
+            let wait_for = scheduler.handle_timeout(worker_queue);
+            match wait_for {
+                Some(timeout) => {
+                    self.condvar.wait_for(&mut scheduler, timeout);
+                }
+                None => self.condvar.wait(&mut scheduler),
+            }
+        }
+    }
+}
+
+
+// For any 'id' (TimerId) in 'self.states', any of following conditions should be false.
+//   1. Exist 'schedule' that contained in 'self.heap', that 'schedule.timer_id == id'
+//   2. '!Arc::ptr_eq(schedules.state_control, self.states[id])'
+//   3. schedules.state_control and 'self.states[id]' both are Wait | WaitRepeating
+// If all of these are met, then the TimerId has been double scheduled.
+struct SchedulerInner {
+    states: HashMap<TimerId, Arc<ScheduleStateControl>>,
+    heap: BinaryHeap<Reverse<TimeOrdered<Schedule>>>,
+    stop: bool,
+}
+impl SchedulerInner {
+    fn new() -> SchedulerInner {
+        SchedulerInner {
+            states: HashMap::new(),
+            heap: BinaryHeap::new(),
+            stop: false,
+        }
+    }
+
+    fn schedule(&mut self, timer_id: TimerId, after: Duration, repeat: Option<Duration>) -> Result<(), ScheduleError> {
+        let state_control = match self.states.entry(timer_id) {
+            Entry::Vacant(entry) => {
+                // The condition 1 is broken. safe
+                let state_control = Arc::new(ScheduleStateControl::new_auto(repeat));
+                entry.insert(Arc::clone(&state_control));
+                state_control
+            }
+            Entry::Occupied(mut entry) => {
+                // The condition 1 met
+                if !entry.get().can_be_removed() {
+                    // Prevent the condition 3
+                    return Err(ScheduleError::TokenAlreadyScheduled)
+                }
+                // The condition 2 met
+                let state_control = Arc::new(ScheduleStateControl::new_auto(repeat));
+                *entry.get_mut() = Arc::clone(&state_control);
+                state_control
+            }
+        };
+
+        let schedule = Schedule {
+            at: Instant::now() + after,
+            timer_id,
+            repeat,
+            state_control,
+        };
+
+        self.heap.push(Reverse(TimeOrdered(schedule)));
+        Ok(())
+    }
+
+    fn reschedule(&mut self, mut schedule: Schedule) -> Result<(), ScheduleError> {
+        schedule.at = Instant::now() + schedule.repeat.expect("Schedule should have repeat interval");
+        match self.states.entry(schedule.timer_id) {
+            Entry::Vacant(entry) => {
+                // The condition 1 is broken. safe
```

Comment:
When does it happen?

Reply:
It happens when repeating schedule was canceled. I found that it should drop schedule since it is canceled. It leads to keep circulating canceled schedule. I fixed that while I'm clarifying condition.